downsampler/src/main/scala/filodb.downsampler/PartitionDownsampler.scala

```diff
@@ -0,0 +1,106 @@
+package filodb.downsampler
+
+import scala.concurrent.duration.FiniteDuration
+
+import com.typesafe.scalalogging.StrictLogging
+import scalaxy.loops._
+
+import filodb.core.downsample._
+import filodb.core.memstore.{TimeSeriesPartition, TimeSeriesShardStats, WriteBufferPool}
+import filodb.core.metadata.Schema
+import filodb.core.store.{AllChunkScan, ReadablePartition}
+import filodb.memory.{BlockMemFactory, MemFactory}
+import filodb.memory.format.SeqRowReader
+
+object PartitionDownsampler extends StrictLogging {
+
+  /**
+    * Creates new donwsample partitions per per the resolutions
+    * * specified by `bufferPools`.
+    * Downsamples all chunks in `partToDownsample` per the resolutions and stoes
+    * downsampled data into the newly created partition.
+    *
+    * @param nativeMemManager downsampled chunks are stored into offheap memory
+    *                         allocated by this memory manager
+    * @param bufferPool buffers for the downsampled partitions are procured from this pool
+    *
+    * @return a TimeSeriesPartition for each resolution to be downsampled.
+    *         NOTE THAT THE PARITIONS NEED TO BE FREED/SHUT DOWN ONCE CHUNKS ARE EXTRACTED FROM THEM
+    */
+  // scalastyle:off method.length parameter.number
+  def downsample(shardStats: TimeSeriesShardStats,
```

Comment:
Yes - lots of params indeed. This will be called a lot of times, just avoided the case class instance to save on the allocation. The number of params problem would stay either way.